File: 2021/09/29/main.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def splitListToParts(self, head: Optional[ListNode], k: int) -> List[Optional[ListNode]]:
        def length(head):
            res = 0
            while head != None:
                res += 1 
                head = head.next
            return res

        i, n = 0, length(head)
        res = []
        subArray = []
        crnt = head
        dd = n // k
        md = n % k
        for i in range(k):
            head = crnt
            for j in range(dd + (i < md) - 1):
                if crnt: 
                    crnt = crnt.next
            if crnt: 
                crnt.next, crnt = None, crnt.next
            res.append(head)

        return res

File: 2021/09/29/main.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func splitListToParts(head *ListNode, k int) []*ListNode {
	n := length(head)
	md := n % k
	res := make([]*ListNode, k)
	i := 0
	for {
		res[i] = head
		i++
		if i == k {
			break
		}

		l := n / k
		if md > 0 {
			l++
			md--
		}

		for l > 1 && head != nil {
			head = head.Next
			l--
		}

		if head == nil {
			break
		}

		head.Next, head = nil, head.Next
	}

	return res
}

func length(head *ListNode) int {
	res := 0
	for head != nil {
		res++
		head = head.Next
	}
	return res
}
